[enocean] Eltako FJ62 support

spledl opened the pull request:
- Add Eltako FJ62 to the list of supported rollershutters
- Add support for Eltako FJ62 rollershutter RPS status response

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

lsiepel reviewed (approved):
Thanks, LGTM. Would be nice if @fruggy83 or @lovery can confirm.

lovery commented:
@spledl Can you explain why you choose to add new eep - F6_00_00?

I can't find such eep in the [enocean specifications](http://tools.enocean-alliance.org/EEPViewer/#1) and on the Eltako website for the device FJ62 I see in the documentation - [1](https://www.eltako.com/fileadmin/downloads/en/_bedienung/FJ62_12-36VDC_30200540-3_gb.pdf) and [2](https://www.eltako.com/fileadmin/downloads/en/_bedienung/FJ62_12-36VDC_30200540-2_gb.pdf) - says F6_10_00, not F6_00_00

spledl commented:
@lovery 

I got the information by tracing the enocean telegrams.

Confirmation telegram of the FJ62 on start down command:
```
2024-06-28 19:51:46.208 [DEBUG] [ernal.transceiver.EnOceanTransceiver] - RADIO_ERP1 with RORG RPS for xxxxxxxx payload F602xxxxxxxx3000FFFFFFFF4300 received
2024-06-28 19:51:46.210 [DEBUG] [rnal.handler.EnOceanBaseThingHandler] - ESP Packet payload F602xxxxxxxx30 for xxxxxxxx received

```
Confirmation telegram of the FJ62 on reaching the lower stop position (rollershutter completeley closed):
```
2024-06-28 19:52:21.035 [DEBUG] [ernal.transceiver.EnOceanTransceiver] - RADIO_ERP1 with RORG RPS for xxxxxxxx payload F650xxxxxxxx3000FFFFFFFF4D00 received
2024-06-28 19:52:21.037 [DEBUG] [rnal.handler.EnOceanBaseThingHandler] - ESP Packet payload F650xxxxxxxx30 for xxxxxxxx received

```
Confirmation telegram of the FJ62 on start up command:
```
2024-06-28 19:53:40.557 [DEBUG] [ernal.transceiver.EnOceanTransceiver] - RADIO_ERP1 with RORG RPS for xxxxxxxx payload F601xxxxxxxx3000FFFFFFFF4600 received
2024-06-28 19:53:40.560 [DEBUG] [rnal.handler.EnOceanBaseThingHandler] - ESP Packet payload F601xxxxxxxx30 for xxxxxxxx received

```
Confirmation telegram of the FJ62 on reaching the upper stop position (rollershutter completeley open):
```
2024-06-28 19:54:15.400 [DEBUG] [ernal.transceiver.EnOceanTransceiver] - RADIO_ERP1 with RORG RPS for xxxxxxxx payload F670xxxxxxxx3000FFFFFFFF4C00 received
2024-06-28 19:54:15.403 [DEBUG] [rnal.handler.EnOceanBaseThingHandler] - ESP Packet payload F670xxxxxxxx30 for xxxxxxxx received

```
see also:
Technical data of the wireless actuators, teach-in list, operating distances and contents of Eltako Wireless telegrams
[https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf](https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf)
Page T-16
CONFIRMATION TELEGRAMS OF BIDIRECTIONAL ACTUATORS.

```
FSB61NP-230V, FSB71, FJ62/12-36V DC, FJ62NP-230V

ORG=         0x05
Data_byte3 = 0x70 = upper stop position,
             0x50 = lower stop position,
             0x01 = Start up, 
             0x02 = Start down

If the actuator is stopped before the end of RV, only the actual elapsed time
is sent indicating the direction in a ORG7 message with the same ID! This is
also the info that the engine has stopped now.

ORG =        0x07
Data_byte3 = driving time in 100 ms MSB
Data_byte2 = driving time in 100 ms LSB
Data_byte1 = 0x01 = driven up or 0x02 = driven down
Data_byte0 = 0x0A (not blocked) 
             or 
             0x0E (blocked)

Remark: The RV time must be set on the device so that the end position is
always reached. If the roller shutter is already at an end position, the relay is
switched on receipt of a drive command anyway (0x01 or 0x02 is sent) and it is
switched off on expiry of the RV. (0x70 or 0x50 is sent).


```
As i don't know what ORG=0x05 means and there was no appropriate EEP in the [enocean specifications](http://tools.enocean-alliance.org/EEPViewer/#1) 
F6-10-00 Window Handle
[http://tools.enocean-alliance.org/EEPViewer/profiles/F6/10/00/F6-10-00.pdf](http://tools.enocean-alliance.org/EEPViewer/profiles/F6/10/00/F6-10-00.pdf)
and
F6-10-01 Window Handle ERP2
[http://tools.enocean-alliance.org/EEPViewer/profiles/F6/10/01/F6-10-01.pdf](http://tools.enocean-alliance.org/EEPViewer/profiles/F6/10/01/F6-10-01.pdf)
doesn't match, therefore i used the unsused EEP F6_00_00.

lovery commented on bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_00/F6_00_00_EltakoFJ62.java:
```diff
@@ -0,0 +1,106 @@
+
+        switch (channelId) {
+            case CHANNEL_ROLLERSHUTTER:
+                switch (bytes[0]) {
```
Have you check what will happen when the values are inverted - like with the CHANNEL_CONTACT?

spledl replied:
In normal mode i get immediately a closed status after a down command --> partially closed.
In normal mode i get a open status after a up command and the shut time of the rollershuter has been fully elapsed.

In inversed mode i get immediately a closed status after a up command --> partially closed.
In inversed mode i get a open status after a down command and the shut time of the rollershuter has been fully elapsed.

lovery commented on bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_00/F6_00_00_EltakoFJ62.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public boolean isValidForTeachIn() {
```
Can you test what will happen with received teachin telegram - will it be associated correctly with the F6_00_00 eep for sending and receiving, or you choose it manually every time?

spledl replied:
I have not tested this.

lovery replied:
Please let me know what happens when you test it?

lovery commented on bundles/org.openhab.binding.enocean/src/main/resources/OH-INF/i18n/enocean.properties:
```diff
@@ -270,16 +270,17 @@ thing-type.config.enocean.rollershutter.enoceanId.description = EnOceanId of dev
 thing-type.config.enocean.rollershutter.pollingInterval.description = Time in seconds after a refresh is triggered, 0 to disable
 thing-type.config.enocean.rollershutter.receivingEEPId.label = EEP for Receiving States
-thing-type.config.enocean.rollershutter.receivingEEPId.option.A5_3F_7F_EltakoFSB = Eltako FSB14/61/71
+thing-type.config.enocean.rollershutter.receivingEEPId.option.A5_3F_7F_EltakoFSB = Eltako FSB14/61/71, FJ62 (4 byte 4BS status)
```
Can you help me understand what is `(4 byte 4BS status)`? Is this for devices made before week 17/21?

spledl replied:
This means that this is a confirmation (status) telegram that is send as
ORG=0x07 -->  4 byte 4BS confirmation telegram
with 4 byte data/payload

see
Technical data of the wireless actuators, teach-in list, operating distances and contents of Eltako Wireless telegrams
[https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf](https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf)
Page T-16
CONFIRMATION TELEGRAMS OF BIDIRECTIONAL ACTUATORS.

FSB61NP-230V, FSB71, FJ62/12-36V DC, FJ62NP-230V

```
...
ORG =        0x07
Data_byte3 = driving time in 100 ms MSB
Data_byte2 = driving time in 100 ms LSB
Data_byte1 = 0x01 = driven up or 0x02 = driven down
Data_byte0 = 0x0A (not blocked) 
             or 
             0x0E (blocked)
...
```

lovery reviewed (commented):
I agree that the standard F6-10-00 doesn't match at all.

If I was making the implementation I would name the eep `F6_10_00_Eltako` or `F6_10_00_EltakoFTKE`, in order to keep the `FUNC` value - 10, and in future if some has problem with its device to be able to find the eep file faster based on the specification of the device where it is said FTKE (EEP F6-10-00)

I don't have a strong opinion about this. I think @fruggy83 knows better how is better to be named.

lovery commented:
> 0x05

Telegram format: 0x05 RPS. 0x06 1BS. 0x07 4BS.

spledl commented on bundles/org.openhab.binding.enocean/README.md:
```diff
@@ -38,7 +38,7 @@ This binding is developed on and tested with the following devices
 - The following Eltako actuators:
   - FSR14 (light switch)
-  - FSB14 (rollershutter)
+  - FSB14/61/71, FJ62 (rollershutter)
```
No i have it tested only with an FJ62NP-230V but the technical specification of Eltako describes that FSB61 and FSB71 this devices uses the same telegrams.
ORG=0x05 --> 1 byte RPS confirmation telegram
ORG=0x07 --> 4 byte 4BS confirmation telegram

see:
Technical data of the wireless actuators, teach-in list, operating distances and contents of Eltako Wireless telegrams
[https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf](https://www.eltako.com/fileadmin/downloads/en/_main_catalogue/Gesamt-Katalog_ChT_gb_highRes.pdf)
Page T-16
CONFIRMATION TELEGRAMS OF BIDIRECTIONAL ACTUATORS.

**_FSB61NP-230V_**, **_FSB71_**, FJ62/12-36V DC, FJ62NP-230V

```
ORG=         0x05
Data_byte3 = 0x70 = upper stop position,
             0x50 = lower stop position,
             0x01 = Start up, 
             0x02 = Start down

If the actuator is stopped before the end of RV, only the actual elapsed time
is sent indicating the direction in a ORG7 message with the same ID! This is
also the info that the engine has stopped now.

ORG =        0x07
Data_byte3 = driving time in 100 ms MSB
Data_byte2 = driving time in 100 ms LSB
Data_byte1 = 0x01 = driven up or 0x02 = driven down
Data_byte0 = 0x0A (not blocked) 
             or 
             0x0E (blocked)

Remark: The RV time must be set on the device so that the end position is
always reached. If the roller shutter is already at an end position, the relay is
switched on receipt of a drive command anyway (0x01 or 0x02 is sent) and it is
switched off on expiry of the RV. (0x70 or 0x50 is sent).

```

lovery replied:
I think you should not add '61/71' here since you didn't test it.

spledl commented:
@lovery 
> I agree that the standard F6-10-00 doesn't match at all.
> 
> If I was making the implementation I would name the eep `F6_10_00_Eltako` or `F6_10_00_EltakoFTKE`, in order to keep the `FUNC` value - 10, and in future if some has problem with its device to be able to find the eep file faster based on the specification of the device where it is said FTKE (EEP F6-10-00)
> 
> I don't have a strong opinion about this. I think @fruggy83 knows better how is better to be named.

The telegram does not come from a mechanical window contact (for which F6_10_00 would be correct) but is generated from the actuator, when the switching time configured in the actuator has expired after a switching operation.

spledl commented on bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPType.java:
```diff
@@ -413,6 +414,9 @@ public enum EEPType {
     CentralCommandBlinds(RORG._4BS, 0x38, 0x08, false, A5_38_08_Blinds.class, THING_TYPE_ROLLERSHUTTER, 0x07,
             CHANNEL_ROLLERSHUTTER, CHANNEL_ANGLE, CHANNEL_TEACHINCMD),
 
+    EltakoFJ62_RPS(RORG.RPS, 0x00, 0x00, false, "EltakoFJ62", 0, F6_00_00_EltakoFJ62.class, THING_TYPE_ROLLERSHUTTER,
```
You are right, but i think this is only relevant, if the EEP F6_00_00 would be used by an other manufacturer in a different way.

lsiepel commented:
@lovery please ping me when you are done with your review. I can only look at the generic code quality and not the data/protocol specifics.

lsiepel commented:
@spedl what happened? Don’t want to merge it anymore?

spledl commented:
> @spledl what happened? Don’t want to merge it anymore?

Sorry but i have accidentally deleted and recreated the merge branch and now i don't know how to fix this problem.

Should i create a new pull request on the new merge branch
[https://github.com/spledl/openhab-addons/tree/EltakoFJ62](https://github.com/spledl/openhab-addons/tree/EltakoFJ62)
or could i otherwise fix this.

btw:
In the current merge branch i have done some changes which where suggested by @lovery.

lsiepel commented:
If you restored the branch you can just re-open this PR. Nothing more should be needed

lsiepel commented:
Then please create a new pull request, and address all questions here or add a self review with the open comments.
